libsql-server/src/http/user/listen.rs

```diff
@@ -0,0 +1,184 @@
+use std::{
+    collections::HashMap,
+    mem,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+use crate::broadcaster::{Action, BroadcastMsg, UpdateSubscription};
+use crate::error::Error;
+use crate::{
+    auth::Authenticated,
+    namespace::{NamespaceName, NamespaceStore},
+};
+use axum::http::Uri;
+use axum::response::{IntoResponse, Response};
+use axum::{body::BoxBody, extract::State as AxumState};
+use axum_extra::{extract::Query, json_lines::JsonLines};
+use futures::{Stream, StreamExt};
+use hyper::HeaderMap;
+use serde::{Deserialize, Serialize};
+
+use super::db_factory::namespace_from_headers;
+use super::AppState;
+
+#[derive(Deserialize)]
+pub struct ListenQuery {
+    table: String,
+    action: Vec<Action>,
+}
+
+pub(super) async fn handle_listen(
+    auth: Authenticated,
+    AxumState(state): AxumState<AppState>,
+    headers: HeaderMap,
+    uri: Uri,
+    mut query: Query<ListenQuery>,
+) -> crate::Result<Response> {
+    let namespace = namespace_from_headers(
+        &headers,
+        state.disable_default_namespace,
+        state.disable_namespaces,
+    )?;
+
+    if !auth.is_namespace_authorized(&namespace) {
+        return Err(Error::NamespaceDoesntExist(namespace.to_string()));
+    }
+
+    if let Some(primary_url) = state.primary_url {
+        return Ok(Response::builder()
+            .status(307)
+            .header("Location", primary_url + uri.path())
+            .body(BoxBody::default())
+            .unwrap());
+    }
+
+    // TODO: validate table
+    let table = mem::take(&mut query.table);
+    let actions = mem::take(&mut query.action);
+
+    let stream =
+        SubscriptionAggregator::new(state.namespaces.clone(), namespace, table, actions).await?;
+    Ok(JsonLines::new(stream).into_response())
+}
+
+static LAGGED_MSG: &str = "some changes were lost";
+
+type AggregatorState = HashMap<Action, u64>;
+
+#[derive(Debug, Serialize)]
+#[serde(rename_all = "snake_case")]
+enum AggregatorEvent {
+    Error(&'static str),
+    #[serde(untagged)]
+    Changes(AggregatorState),
+}
+
+type AggregatorResult = Result<AggregatorEvent, Error>;
+type AggregatorPoll = Poll<Option<AggregatorResult>>;
+
+struct SubscriptionAggregator {
+    actions: Vec<Action>,
+    subscription: UpdateSubscription,
+    state: AggregatorState,
+    store: NamespaceStore,
+    namespace: NamespaceName,
+    table: String,
+    errored: bool,
+}
```

Comment:
I wonder: why don't you aggregate at the source? That would make things so much more efficient...